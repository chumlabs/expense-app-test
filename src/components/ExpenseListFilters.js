// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
// - components
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filterActions';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    // console.log(e.target.value);
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <h3>Applied Filters</h3>
        Text search:
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
        Sort expenses by:
        <select
          name="sort"
          id="sort"
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">date</option>
          <option value="amount">amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          startDateId={'start'}
          endDateId={'end'}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: date => dispatch(setStartDate(date)),
  setEndDate: date => dispatch(setEndDate(date)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
