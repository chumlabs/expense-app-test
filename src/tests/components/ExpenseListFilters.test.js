import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { ExpenseListFilters } from '../../components/ExpenseListFilters.js';
import { defaultFilters, uniqueFilters } from '../fixtures/filterData'; // mock filters data

/* ExpenseListFilter props:
  filters
  setStartDate
  setEndDate
  setTextFilter
  sortByDate
  sortByAmount
*/

// test setup
let setStartDate, setEndDate, setTextFilter, sortByDate, sortByAmount, wrapper;

beforeEach(() => {
  setStartDate = jest.fn();
  setEndDate = jest.fn();
  setTextFilter = jest.fn();
  sortByDate = jest.fn();
  sortByAmount = jest.fn();
  wrapper = shallow(
    <ExpenseListFilters
      filters={defaultFilters}
      setStartDate={setStartDate}
      setEndDate={setEndDate}
      setTextFilter={setTextFilter}
      sortByDate={sortByDate}
      sortByAmount={sortByAmount}
    />
  );
});

// tests
test('should render ExpenseListFilters component properly (with default filters)', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseListFilters component properly (with unique filters)', () => {
  wrapper.setProps({ filters: uniqueFilters });
  expect(wrapper).toMatchSnapshot();
});

test('should handle text filter change', () => {
  const value = 'more';
  wrapper
    .find('input')
    .at(0)
    .simulate('change', { target: { value } });
  expect(setTextFilter).toHaveBeenLastCalledWith(value);
});

test('should sort by date', () => {
  const value = 'date';
  wrapper.setProps({ filters: uniqueFilters }); // to set text filter to non-default value
  wrapper.find('select').simulate('change', { target: { value } });
  expect(sortByDate).toHaveBeenCalled();
});

test('should sort by amount', () => {
  const value = 'amount';
  wrapper.find('select').simulate('change', { target: { value } });
  expect(sortByAmount).toHaveBeenCalled();
});

test('should handle start & end date change', () => {
  const startDate = moment(0).add(2, 'years');
  const endDate = moment(0).add(25, 'years');
  wrapper.find('[onDatesChange]').prop('onDatesChange')({ startDate, endDate });
  expect(setStartDate).toHaveBeenLastCalledWith(startDate);
  expect(setEndDate).toHaveBeenLastCalledWith(endDate);
});

test('should handle date focus change', () => {
  const calendarFocused = 'startDate';
  wrapper.find('[onFocusChange]').prop('onFocusChange')(calendarFocused);
  expect(wrapper.state('calendarFocused')).toBe(calendarFocused);
});
