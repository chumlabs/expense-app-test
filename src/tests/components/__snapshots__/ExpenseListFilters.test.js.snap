// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters component properly (with default filters) 1`] = `
<div>
  <h3>
    Applied Filters
  </h3>
  Text search:
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  Sort expenses by:
  <select
    id="sort"
    name="sort"
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      date
    </option>
    <option
      value="amount"
    >
      amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    block={false}
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    disabled={false}
    enableOutsideDays={false}
    endDate={null}
    endDateId="end"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={null}
    startDateId="start"
    verticalHeight={null}
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;

exports[`should render ExpenseListFilters component properly (with unique filters) 1`] = `
<div>
  <h3>
    Applied Filters
  </h3>
  Text search:
  <input
    onChange={[Function]}
    type="text"
    value="duck"
  />
  Sort expenses by:
  <select
    id="sort"
    name="sort"
    onChange={[Function]}
    value="amount"
  >
    <option
      value="date"
    >
      date
    </option>
    <option
      value="amount"
    >
      amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    block={false}
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    disabled={false}
    enableOutsideDays={false}
    endDate={"2017-05-01T16:00:00.000Z"}
    endDateId="end"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={"2017-04-01T16:00:00.000Z"}
    startDateId="start"
    verticalHeight={null}
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
